### PR TITLE
Add --render-endpoint to prepare_data.py for vLLM-delegated tokenization (#652)

### PR DESCRIPTION
**Pull request by WindChimeRan**
## Purpose

Adds a `--render-endpoint` flag to `prepare_data.py` that delegates tokenization to a running vLLM render server instead of doing it locally with HF `apply_chat_template`. 

The --render-endpoint should point at the same vllm serve instance used for hidden-state extraction. The render endpoint runs on the same server with zero GPU overhead. A separate vllm launch render instance is only needed for GPU-less offline preprocessing.

vLLM prerequisite PR has been merged, bun not released yet: https://github.com/vllm-project/vllm/pull/46846

When `--render-endpoint` is set, raw conversations are sent to `POST /v1/chat/completions/render` and the returned `token_ids` + `loss_mask` are used directly, eliminating the need for local chat-template rendering, assistant-pattern detection, and regex-based loss masking. 

Motivation: online training already requires a vLLM instance, so tokenization should happen once on the vLLM side rather than being duplicated between HF and vLLM. See RFC #652 for discussion.

## Tokenization 

* Render + server mask (template has `{% generation %}`): token_ids and mask both from /render → 1 tokenization, server-side, no local re-tokenization.
* Render + regex fallback (no gen-tags): server token_ids kept as input_ids; mask built from one local decode → re-tokenize for offsets only (row dropped if it doesn't round-trip) → 2 tokenizations (1 server + 1 local, offsets-only).
* Default HF path (no --render-endpoint, unchanged): single local apply_chat_template → 1 tokenization, no server.

## Corner cases

The current renderer endpoint + fallback may not fully guarantee the correctness of some corner cases, 

e.g., #659 , #645, #690 need to investigate further

## Tests

* passed end2end tests using vllm PR: https://github.com/vllm-project/vllm/pull/46846

## Limitation:
* **[Future work]** No concurrency: the render path is a sequential for loop of blocking HTTP calls.
* The headline benefit is niche. For the common models (Qwen2.5/3/VL — no {% generation %} tags), the render path falls to the same local regex the default path uses. 

## Checklist

I have filled in:

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan/results, such as providing test command and pasting the results.
- [ ] (Optional) The necessary documentation update.
- [x] I (a human) have written or reviewed the code in this pr to the best of my ability.

### CODE REVIEW

File: scripts/prepare_data.py [hunk 1]
```diff
@@ -105,6 +105,18 @@ def parse_args():
             "per conversation for data augmentation."
         ),
     )
+    parser.add_argument(
+        "--render-endpoint",
```
Comment (rahul-tuli) [inline scripts/prepare_data.py]: nit: Name is acceptable and the help text makes it clear, but `--render-endpoint` might invite users to put in the complete endpoint `/v1/chat/completions/render` should we consider renaming to something like `--render-server-url` or `--vllm-server-url`?

File: src/speculators/data_generation/preprocessing.py [hunk 2]
```diff
@@ -773,18 +883,29 @@ def load_and_preprocess_dataset(
...
-            dataset=raw_dataset,
-            processor=processor,
-            max_length=seq_length,
-            num_proc=build_dataset_num_proc,
-            assistant_pattern=assistant_pattern,
-            turn_dropout=turn_dropout,
-            minimum_valid_tokens=minimum_valid_tokens,
-        )
+        if render_endpoint is not None:
+            preprocessed_dataset = build_dataset_from_render(
```
Comment (rahul-tuli) [inline src/speculators/data_generation/preprocessing.py:1026]: Could we warn/raise if both render-endpoint and assisstant pattern are set?

File: src/speculators/data_generation/preprocessing.py [hunk 3]
```diff
@@ -709,32 +837,10 @@ def load_and_preprocess_dataset(
     turn_dropout: bool = False,
     minimum_valid_tokens: int | None = None,
     trust_remote_code: bool = False,
+    render_endpoint: str | None = None,
+    chat_template_kwargs: dict | None = None,
```
Comment (rahul-tuli) [inline src/speculators/data_generation/preprocessing.py:936]: Shouldn't we also add `chat_template_kwargs` in the CLI?

File: src/speculators/data_generation/preprocessing.py [hunk 4]
```diff
@@ -709,32 +837,10 @@ def load_and_preprocess_dataset(
...
-                          chat template.
-        turn_dropout: If True, randomly keeps first N consecutive turns per
-                     conversation
-        minimum_valid_tokens: Number of tokens to consider for a valid sample
-        trust_remote_code: If True, allows executing code from HF Hub.
-
-    Returns:
-        Tuple of (preprocessed_dataset, processor)
-    """
+    """Load, tokenize, and preprocess a dataset for EAGLE3 training."""
```
Comment (rahul-tuli) [inline src/speculators/data_generation/preprocessing.py]: Since this is a public-ish function, could we update the docstring with new additions instead of truncating it?

**Review by rahul-tuli — COMMENTED** [review]
Thank you for working on this @WindChimeRan, the diff looks good to me pending a few small comments, could you respond and update with latest main, and I will approve the workflows to run, great job on this!

File: tests/integration/datagen/test_render_multimodal.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+from datasets import Dataset as HFDataset
+from PIL import Image
+
+from speculators.data_generation.preprocessing import (
+    build_dataset_from_render,
+    load_processor,
+)
+
+VLLM_PYTHON = "/home/ran/workspace/vllm/.venv/bin/python"
```
Comment (shanjiaz) [inline tests/integration/datagen/test_render_multimodal.py]: probably needs to fix this.

File: src/speculators/data_generation/render_client.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
+"""Client for vLLM's /v1/chat/completions/render endpoint (RFC #652)."""
+
+from http import HTTPStatus
+
+import httpx
```
Comment (shanjiaz) [inline src/speculators/data_generation/render_client.py:5]: Is httpx in our requirements, could you check.

Comment (rahul-tuli) [reply src/speculators/data_generation/render_client.py:5]: I think it's installed by openai dependency

Comment (WindChimeRan) [reply src/speculators/data_generation/render_client.py:5]: added the dependency explicitly.

File: scripts/prepare_data.py (hunk 1 above)
Comment (WindChimeRan) [reply scripts/prepare_data.py]: use `--vllm-server-url` now

**Comment by WindChimeRan** [comment]
Thanks for the review! All addressed @rahul-tuli @shanjiaz

File: tests/integration/datagen/test_render_multimodal.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+from speculators.data_generation.preprocessing import (
+    build_dataset_from_render,
+    load_processor,
+)
+from tests.e2e.utils import VLLM_PYTHON
+
+MODEL = "Qwen/Qwen3-VL-2B-Instruct"
+MEDIA_DIR = "/tmp"
+IMG_PATH = "/tmp/speculators_render_mm_test.png"
```
Comment (rahul-tuli) [inline tests/integration/datagen/test_render_multimodal.py]: Instead of hardcoding these temp paths, I believe it would be nicer if we can use pytest primitives or context managers for these?

You could use something like the following maybe?
```python
@pytest.fixture(scope="module")
def test_dir(tmp_path_factory) -> Path:
    """Provides a module-scoped isolated temporary directory."""
    return tmp_path_factory.mktemp("speculators_render")


@pytest.fixture(scope="module")
def image(test_dir: Path) -> str:
    img_path = test_dir / "render_mm_test.png"
    arr = np.zeros((128, 128, 3), dtype=np.uint8)
    arr[:64, :, 0] = 200
    arr[64:, :, 2] = 200
    arr[32:96, 32:96, 1] = 180
    Image.fromarray(arr).save(img_path)
    return str(img_path)


@pytest.fixture(scope="module")
def render_server(test_dir: Path) -> str:
    ...
```

**Comment by rahul-tuli** [comment]
LGTM, pending CI!

**Review by rahul-tuli — APPROVED** [review]

**Comment by WindChimeRan** [comment]
this has design conflicts with #774

**Comment by WindChimeRan** [comment]
design conflict with #794
